Amazon.QLDB.Driver/retry/AsyncRetryHandler.cs

```diff
@@ -16,18 +16,89 @@ namespace Amazon.QLDB.Driver
     using System;
     using System.Threading;
     using System.Threading.Tasks;
+    using Amazon.QLDBSession.Model;
+    using Microsoft.Extensions.Logging;
 
-    internal class AsyncRetryHandler : IAsyncRetryHandler
+    /// <summary>
+    /// <para>The asynchronous implementation of Retry Handler.</para>
+    ///
+    /// <para>The driver retries in two scenarios: retrying inside a session, and retrying with another session. In the second case,
+    /// it would require a <i>recover</i> action to reset the session into a working state.
+    /// </summary>
+    internal class AsyncRetryHandler : BaseRetryHandler, IAsyncRetryHandler
     {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AsyncRetryHandler"/> class.
+        /// </summary>
+        /// <param name="logger">The logger to record retries.</param>
+        public AsyncRetryHandler(ILogger logger)
+            : base(logger)
+        {
+        }
+
+        /// <inheritdoc/>
         public async Task<T> RetriableExecute<T>(
-            Func<Task<T>> func,
+            Func<CancellationToken, Task<T>> func,
             RetryPolicy retryPolicy,
-            Func<Task> newSessionAction,
-            Func<Task> nextSessionAction,
-            Action<int> retryAction,
+            Func<CancellationToken, Task> newSessionAction,
+            Func<CancellationToken, Task> nextSessionAction,
+            Func<int, CancellationToken, Task> retryAction,
             CancellationToken cancellationToken = default)
         {
-            throw new NotImplementedException();
+            Exception last = null;
+            int retryAttempt = 0;
+
+            while (true)
+            {
+                try
+                {
+                    return await func(cancellationToken);
+                }
+                catch (QldbTransactionException ex)
+                {
+                    var iex = ex.InnerException != null ? ex.InnerException : ex;
+
+                    if (!(ex is RetriableException))
+                    {
+                        throw iex;
+                    }
+
+                    last = iex;
```

Comment:
Let's just throw IEX at the bottom. Don't need to instantiate last nor assign iex to last here.

Reply:
Done.